lib/cyrusdb_twom.c

```diff
@@ -0,0 +1,2597 @@
+{
+    struct copy_rock *cr = (struct copy_rock *)rock;
+    int i;
+
+    /* minimal logic from find_loc and stitch knowing that we're
+     * always writing at the end of a file */
+    struct tm_loc *loc = &cr->db->loc;
+    const char *ptr = LOCPTR(loc);
+    uint8_t level = LEVEL(ptr);
+    for (i = 0; i < level; i++)
+        loc->backloc[i] = loc->offset;
+    loc->is_exactmatch = 0;
+    return store_here(cr->db, key, keylen, data, datalen);
+}
+
+static int replay_cb(void *rock,
+                     const char *key, size_t keylen,
+                     const char *data, size_t datalen)
+{
+    struct copy_rock *cr = (struct copy_rock *)rock;
+    return skipwrite(cr->db, key, keylen, data, datalen, 1);
+}
+
+static int mm_rename(struct dbengine *new, struct dbengine *db)
+{
+    char *copy = xstrdup(db->fname);
+    const char *dir = dirname(copy);
+    int r = 0;
+
+    #if defined(O_DIRECTORY)
+    int dirfd = open(dir, O_RDONLY|O_DIRECTORY, 0600);
+#else
+    int dirfd = open(dir, O_RDONLY, 0600);
+#endif
+    if (dirfd < 0) {
+        xsyslog(LOG_ERR, "IOERROR: open directory failed",
+                         "filename=<%s> newname=<%s> directory=<%s>",
+                         db->fname, new->fname, dir);
+        r = dirfd;
+        goto done;
+    }
+
+    r = rename(new->fname, db->fname);
+    if (r) goto done;
+
+    if (fsync(dirfd) < 0) {
+        xsyslog(LOG_ERR, "IOERROR: fsync directory failed",
+                         "filename=<%s> newname=<%s> directory=<%s>",
+                         db->fname, new->fname, dir);
+        // but we can't abort now, we've already renamed, so we need to
+        // carry on and update our object
+    }
+
+    // now that the rename is done, we can move everything over
+    unlock(db, NULL);
+    char *fname = db->fname;
+    empty_db(db);
+    *db = *new;
+    free(db->fname);
+    db->fname = fname;
+    free(new);
+
+done:
+    if (dirfd >= 0) close(dirfd);
+    free(copy);
+    return r;
+}
+
+static int mycheckpoint(struct dbengine *db)
+{
+    char newfname[1024];
+    clock_t start = sclock();
+    struct copy_rock cr;
+    int r = 0;
+    struct txn *txn = NULL;
+
+    r = refreshtxn(db, 1/*shared*/, &txn);
+    if (r) return r;
+
+    size_t old_size = db->openfile->header.current_size;
+
+    r = myconsistent(db, txn);
+    if (r) {
+        syslog(LOG_ERR, "db %s, inconsistent pre-checkpoint, bailing out",
+               db->fname);
+        int r2 = mycommit(db, txn);
+        return r2 ? r2 : r;
+    }
+
+    /* open fname.NEW */
+    snprintf(newfname, sizeof(newfname), "%s.NEW", db->fname);
+
+    cr.db = NULL;
+    cr.tid = NULL;
+    r = opendb(newfname, CYRUSDB_CREATE, &cr.db, &cr.tid);
+    if (r) return r;
+
+    // has it already repacked?
+    if (cr.db->openfile->header.num_records ||
+        cr.db->openfile->header.generation > 1 ||
+        !memcmp(cr.db->openfile->header.uuid, db->openfile->header.uuid, 16)) {
+        r = myabort(cr.db, cr.tid);
```

Comment:
IMO with long conditions it's easier to see their shape and where they end at a glance if they're formatted like this:

```suggestion
    if (cr.db->openfile->header.num_records
        || cr.db->openfile->header.generation > 1
        || !memcmp(cr.db->openfile->header.uuid, db->openfile->header.uuid, 16))
    {
        r = myabort(cr.db, cr.tid);
```

Reply:
Sure, I've reordered the tests too, looks a bit neater and speed will be similar